File: v12/src/worker_manager/schema.mjs
// The Worker Manager's OWN durable control store schema.
//
// W2929. This is not the authority and it is not a migration of one: the
// authority owns Work state, generations, gates, claim outcomes and the
// workflow receipts, and this store owns the facts a manager is
// authoritative for — which offers it issued, which runtime attempts it
// made, what its adapters observed, and how its own operations settled.
//
// THE SPLIT IS THE POINT. `SPEC.md` §3 of W151 says a manager-local or
// sidecar identity is non-conforming: a manager that stored a claim or a
// generation would be a second authority, and two authorities disagreeing
// is not a bug that can be fixed afterwards. So there is no `claim` here,
// no generation counter and no Work phase — only the fixed claim OPERATION
// identity the authority settles, recorded so a restart can go and ask.
//
// WHAT MAY NEVER BE HERE. Bearer tokens, credentials, raw provider
// approval payloads, and the authority's configuration or store locator.
// Only verifier and body digests, redacted diagnostics, and identities.
// A regression walks every column of every table asserting the bearer is
// absent, because "we do not write it" is a property of code and this is a
// property of the file.

// VERSION HISTORY, one line each, because an incompatible binary refuses on
// this number and the reason it moved is what tells a reader whether their
// store is behind:
//
//   2: the assignment bindings item 3's review required — the generation a
//      claim actually produced, on the offer, and the full four-part
//      assignment on the attempt. A manager that stored three of four
//      fields could not compare the fourth.
//   3: one claimed offer per runtime attempt, as a UNIQUE INDEX. The
//      allocator already behaved; the store did not say so.
//   4: the frozen output — the immutable result record and its artifact
//      references.
//   5: the RETAINED MANIFESTS. A digest is not a record: the freeze review
//      found that neither the input declaration nor the sealed result
//      survived the call that produced it.
//   6: the TRUSTED INTAKE DECISION, with its retention separate from it.
//   7: the intake decision BOUND TO THE MATERIAL it decided. A locator is
//      where something is; it is not which immutable result was judged.
//   8: the CERTIFIED PROFILE BYTES, and the agent sessions opened under
//      them. A profile digest cannot be read for the policy it pins.
//   9: AT MOST ONE OPEN SESSION PER POSTURE, as a partial unique index.
//      Freshness and concurrency are two rules and were tested as one.
//  10: TURNS, with the manager deadline every one of them has and the
//      dispositions its outcome permits.
//  11: the turn's own CANONICAL DOCUMENT, retained beside the summary. A
//      row that cannot represent the frozen record is not the record.
export const SCHEMA_VERSION = 11;

export const SCHEMA = `
PRAGMA journal_mode = WAL;
PRAGMA foreign_keys = ON;
-- Concurrent managers WAIT for the write lock rather than failing with
-- SQLITE_BUSY. The one-offer-per-Work rule must be decided by a refusal
-- INSIDE a transaction; losing by not getting a transaction at all is an
-- error, and the two are not interchangeable.
PRAGMA busy_timeout = 5000;

CREATE TABLE IF NOT EXISTS meta (
	key   TEXT PRIMARY KEY,
	value TEXT NOT NULL
) STRICT;

-- The certified profiles this manager may act under. A profile is
-- certified by DIGEST: "the runtime profile we agreed on" is a byte
-- identity, not a name, or a later edit to a file would silently
-- recertify itself.
CREATE TABLE IF NOT EXISTS profiles (
	kind        TEXT NOT NULL CHECK (kind IN ('runtime', 'agent-session')),
	name        TEXT NOT NULL,
	digest      TEXT NOT NULL,
	-- THE CERTIFIED BYTES, for a kind this manager certifies from a document
	-- rather than from a digest somebody hands it. A digest cannot be read
	-- for the per-posture policy a session must pin, which is the same
	-- lesson the freeze review taught about results: a digest is not a
	-- record. Null for a runtime profile, whose document W2930 owns.
	body        TEXT,
	certified_at TEXT NOT NULL,
	withdrawn_at TEXT,
	PRIMARY KEY (kind, name)
) STRICT;

-- ONE AGENT SESSION per (attempt, posture, epoch).
--
-- Consent and execution NEVER share an epoch or a connection, and an epoch is
-- FRESH every time: the manager never resumes, forks or promotes one, so the
-- key is a fact rather than a convention. The reference labels evidence; it
-- is never an assignment identity and it authorizes nothing.
CREATE TABLE IF NOT EXISTS agent_sessions (
	runtime_attempt_id TEXT NOT NULL REFERENCES attempts(runtime_attempt_id),
	posture            TEXT NOT NULL
	                   CHECK (posture IN ('consent', 'execution')),
	session_epoch      INTEGER NOT NULL CHECK (session_epoch >= 1),
	profile_digest     TEXT NOT NULL,
	-- the per-posture policy this session pinned, by digest of the exact
	-- policy object the certified profile carries for this posture
	pinned_policy      TEXT NOT NULL,
	work_id            TEXT NOT NULL,
	authority_uuid     TEXT NOT NULL,
	-- exactly null for a consent session, which exists before any claim, and
	-- exactly the assignment for an execution one
	participant        TEXT,
	generation         INTEGER,
	provider_session_id TEXT,
	state              TEXT NOT NULL,
	opened_at          TEXT NOT NULL,
	PRIMARY KEY (runtime_attempt_id, posture, session_epoch)
) STRICT;

-- AT MOST ONE OPEN SESSION PER POSTURE, across every manager process.
--
-- Review [P1]: freshness and concurrency are two rules and the first version
-- tested them as one — allocating MAX+1 and inserting unconditionally, so a
-- posture could hold three simultaneously open epochs. A partial unique index
-- rather than a read-then-write check, because two managers racing on
-- separate connections both pass any read and only the database can refuse
-- the second.
--
-- 'closed' is the only state that frees the posture. 'unknown' deliberately
-- does NOT: transport ambiguity is where a second session would be most
-- tempting and least safe, and re-identification after it is a gate this
-- slice has not built.
--
-- (Plain quotes, not backticks: this whole schema is a JS template literal
-- and a backticked word terminates it. THIRD occurrence in this file.)
CREATE UNIQUE INDEX IF NOT EXISTS agent_sessions_one_open_per_posture
	ON agent_sessions (runtime_attempt_id, posture)
	WHERE state <> 'closed';

-- ONE TURN, inside one session epoch.
--
-- EVERY TURN HAS A MANAGER DEADLINE, so the column is NOT NULL: a turn whose
-- deadline is optional is a turn that can wait forever, and 'timeout' is one
-- of the two honest outcomes precisely because the deadline is always there.
--
-- 'permitted' is DERIVED from the outcome and stored so a later reader sees
-- the gate that was applied rather than re-deriving one from a table that may
-- have moved. The turn outcome GATES the worker disposition and never
-- chooses it.
CREATE TABLE IF NOT EXISTS turns (
	turn_id            TEXT PRIMARY KEY,
	runtime_attempt_id TEXT NOT NULL REFERENCES attempts(runtime_attempt_id),
	posture            TEXT NOT NULL,
	session_epoch      INTEGER NOT NULL,
	prompt_digest      TEXT NOT NULL,
	started_at         TEXT NOT NULL,
	deadline_at        TEXT NOT NULL,
	ended_at           TEXT NOT NULL,
	outcome            TEXT NOT NULL,
	-- the exact provider observation the outcome was derived from; 'none' is
	-- how a timeout or a lost transport says it has nothing
	terminal_kind      TEXT NOT NULL,
	terminal_value     TEXT,
	conclusive         INTEGER NOT NULL,
	permitted          TEXT NOT NULL,
	event_count        INTEGER NOT NULL DEFAULT 0,
	late_event_count   INTEGER NOT NULL DEFAULT 0,
	dropped_event_count INTEGER NOT NULL DEFAULT 0,
	dropped_event_bytes INTEGER NOT NULL DEFAULT 0,
	-- THE EVIDENCE THAT SELECTED THE OUTCOME, beside the outcome. A durable
	-- 'policy-failed' whose failure list vanished is not the record that
	-- outcome came from, and a reader should not have to open the sealed
	-- document to ask which violation ended the turn.
	policy_failures    TEXT NOT NULL DEFAULT '[]',
	-- the sealed frozen turnRecord, RFC 8785 canonical. The columns above
	-- are the queryable summary; this is the record. A summary that cannot
	-- represent policy failures, evidence, diagnostics or the seal is not
	-- what was accepted, and reproducing what was accepted is the point.
	body               TEXT NOT NULL,
	document_digest    TEXT NOT NULL,
	recorded_at        TEXT NOT NULL,
	FOREIGN KEY (runtime_attempt_id, posture, session_epoch)
		REFERENCES agent_sessions (runtime_attempt_id, posture, session_epoch)
) STRICT;

-- One offer. The verifier is stored; the bearer never is.
CREATE TABLE IF NOT EXISTS offers (
	offer_id           TEXT PRIMARY KEY,
	work_id            TEXT NOT NULL,
	authority_uuid     TEXT NOT NULL,
	participant        TEXT NOT NULL,
	runtime_attempt_id TEXT NOT NULL,
	incarnation        TEXT NOT NULL,
	-- advisory evidence only: the readiness episode this offer answered.
	-- It is NOT the assignment episode and grants nothing.
	readiness_episode  INTEGER,
	input_digest       TEXT NOT NULL,
	policy_digest      TEXT NOT NULL,
	profile_digest     TEXT NOT NULL,
	-- "sha256:<hex over the bearer's own UTF-8 bytes>" — W151 §7's ONE
	-- derivation, pinned by the W4487 re-review. Never the bearer.
	verifier           TEXT NOT NULL,
	verifier_spent     INTEGER NOT NULL DEFAULT 0,
	issued_at          TEXT NOT NULL,
	expires_at         TEXT NOT NULL,
	-- W151's exact offer vocabulary. 'issued' and 'accepted' are the two
	-- NONTERMINAL states, which is what the partial index below counts.
	state              TEXT NOT NULL CHECK (state IN (
		'issued', 'accepted', 'claimed', 'declined', 'expired',
		'settlement-expired', 'claim-refused', 'abandoned-after-restart')),
	-- the durable decision record: its prose is an operand of the
	-- operation signature, so it is stored beside the state it caused
	decision_reason    TEXT,
	decided_at         TEXT,
	-- acceptance's separate facts. The intent is immutable once written,
	-- and the settlement deadline is a DIFFERENT deadline from expiry.
	intent_digest      TEXT,
	accepted_at        TEXT,
	settle_by          TEXT,
	claim_operation_id TEXT,
	claim_signature    TEXT,
	-- the generation the claim COMMITTED, recorded when it settled. It is
	-- what an attempt's activation compares against: a live assignment
	-- somewhere in the authority is not proof that THIS offer claimed it.
	claim_generation   INTEGER
) STRICT;

-- ONE OFFER PER RUNTIME ATTEMPT, at the database boundary.
--
-- Review [P1]: activation asks for THIS attempt's claim, and that question
-- has no honest answer if two claimed rows can share an attempt identity —
-- 'SELECT ... LIMIT 1' would merely pick one by unspecified row order. I
-- argued this was unwitnessable by construction; the construction was a
-- property of the ALLOCATOR, not of the store, and the store is what
-- activation reads. An invariant only the writer maintains is not an
-- invariant.
CREATE UNIQUE INDEX IF NOT EXISTS offers_one_per_attempt
	ON offers (runtime_attempt_id);

-- AT MOST ONE nonterminal offer per Work, across every manager process.
-- A partial unique index rather than a check in code: two managers racing
-- on separate connections both pass any read-then-write check, and only
-- the database can refuse the second.
CREATE UNIQUE INDEX IF NOT EXISTS offers_one_live_per_work
	ON offers (authority_uuid, work_id)
	WHERE state IN ('issued', 'accepted');

-- One runtime attempt. The ten orthogonal axes of the runtime-attempt
-- manifest are columns rather than a JSON blob, so a regression can ask
-- the database whether an axis regressed.
CREATE TABLE IF NOT EXISTS attempts (
	runtime_attempt_id TEXT PRIMARY KEY,
	work_id            TEXT,
	authority_uuid     TEXT,
	-- the exact live assignment this attempt belongs to, once there is
	-- one. Null before activation: an attempt exists from the offer.
	-- ALL FOUR FIELDS, because an assignment is a four-part identity and
	-- comparing three of them is comparing a different thing.
	assignment_generation INTEGER,
	assignment_participant TEXT,
	adapter_name       TEXT NOT NULL,
	adapter_digest     TEXT NOT NULL,
	profile_digest     TEXT NOT NULL,
	input_digest       TEXT,
	policy_digest      TEXT,
	image_digest       TEXT,
	toolchain_digest   TEXT,
	-- opaque and adapter-minted; it carries no authority
	runtime_id         TEXT,
	observation_seq    INTEGER NOT NULL DEFAULT 0,
	consent_runtime    TEXT NOT NULL DEFAULT 'not-started',
	execution_runtime  TEXT NOT NULL DEFAULT 'not-started',
	output             TEXT NOT NULL DEFAULT 'open',
	worker_disposition TEXT NOT NULL DEFAULT 'none',
	proposal           TEXT NOT NULL DEFAULT 'none',
	verification       TEXT NOT NULL DEFAULT 'none',
	technical_review   TEXT NOT NULL DEFAULT 'none',
	approval           TEXT NOT NULL DEFAULT 'none',
	integration        TEXT NOT NULL DEFAULT 'none',
	cleanup            TEXT NOT NULL DEFAULT 'pending',
	created_at         TEXT NOT NULL,
	observed_at        TEXT
) STRICT;

-- The MANAGER's own operation journal, separate from the authority's.
-- §4.2: success at one boundary does not imply success at the other, and
-- reconciliation queries both exact records.
CREATE TABLE IF NOT EXISTS operations (
	operation_id TEXT PRIMARY KEY,
	kind         TEXT NOT NULL,
	signature    TEXT NOT NULL,
	state        TEXT NOT NULL
	             CHECK (state IN ('committed', 'refused')),
	-- byte-stable: an exact retry returns THIS, not a recomputation that
	-- might differ in member order
	result       TEXT,
	-- Review 2026-08-22 [P1]: this held only a MESSAGE, and replay then
	-- rebuilt every durable refusal as 'refused.precondition'. A durable
	-- 'policy.retention' came back with a different portable meaning and a
	-- different retry policy — the decision ran once, but the retry was not
	-- a replay of the same refusal. The whole sealed outcome is stored, so
	-- the first answer is REPRODUCED rather than re-derived.
	refusal      TEXT,
	settled_at   TEXT NOT NULL
) STRICT;

-- THE TRUSTED INTAKE DECISION.
--
-- Not one of the frozen axes, on purpose: the ten axes are the MANAGER's own
-- observations, and this is somebody else's decision that the manager merely
-- records. Nothing here reaches the authority.
--
-- RETENTION IS NOT ACCEPTANCE (SPEC 6.4). Whether intake wanted the material
-- and where the material went are two facts, so they are two columns; a
-- rejected draft that is retained under policy is an ordinary outcome and
-- collapsing them would make it unsayable.
--
-- The assignment is stored with the decision because a decision read under a
-- different generation is a decision about something else.
CREATE TABLE IF NOT EXISTS intake (
	runtime_attempt_id TEXT PRIMARY KEY
	                   REFERENCES attempts(runtime_attempt_id),
	-- WHICH MATERIAL WAS JUDGED. Review [P1]: the decision stored only a
	-- locator, so a restart could not prove which immutable result it
	-- concerned, and an attempt with no sealed output at all could be
	-- accepted or rejected. A locator is where something is.
	result_digest      TEXT NOT NULL REFERENCES manifests(digest),
	disposition        TEXT NOT NULL
	                   CHECK (disposition IN ('accepted', 'rejected')),
	retention          TEXT NOT NULL
	                   CHECK (retention IN ('retained', 'quarantined')),
	locator            TEXT NOT NULL,
	retain_until       TEXT,
	reason             TEXT,
	work_id            TEXT NOT NULL,
	authority_uuid     TEXT NOT NULL,
	participant        TEXT NOT NULL,
	generation         INTEGER NOT NULL,
	decided_at         TEXT NOT NULL
) STRICT;

-- Adapter observations, keyed by their FULL source scope.
--
-- Review 2026-08-22 [P1]: the key was '(runtime_attempt_id, source_seq)',
-- and the frozen contract scopes 'source_seq' to one adapter INCARNATION.
-- So an adapter restart — a valid new incarnation whose sequence begins
-- again at 1 — collided with the previous incarnation's sequence 1, and the
-- required restart path became indistinguishable from a conflicting
-- duplicate before any monotonic logic ran.
--
-- The source identity and the manager's own ordering are two different
-- facts and are stored as two: this key answers "have I seen THIS
-- observation", and 'manager_seq' carries the per-attempt monotonicity that
-- axis logic enforces separately.
CREATE TABLE IF NOT EXISTS observations (
	runtime_attempt_id TEXT NOT NULL REFERENCES attempts(runtime_attempt_id),
	incarnation        TEXT NOT NULL,
	source_seq         INTEGER NOT NULL,
	runtime_id         TEXT,
	-- the immutable digest of what was observed, which is what makes
	-- "the same observation again" answerable
	observation_digest TEXT NOT NULL,
	manager_seq        INTEGER NOT NULL,
	observed_at        TEXT NOT NULL,
	PRIMARY KEY (runtime_attempt_id, incarnation, source_seq)
) STRICT;
-- The manager's own sequence is unique PER ATTEMPT regardless of which
-- incarnation produced the observation: it is this manager's ordering, not
-- the adapter's.
CREATE UNIQUE INDEX IF NOT EXISTS observations_manager_seq
	ON observations (runtime_attempt_id, manager_seq);

-- THE FROZEN OUTPUT. Store surface item 7: an immutable freeze/result
-- record, its artifact references and their exact digests.
--
-- One row per attempt, and the PRIMARY KEY says so at the database
-- boundary. The manager operation journal decides replay-versus-collision
-- for the product path; this is what a store written some other way still
-- cannot violate.
-- THE RETAINED MANIFESTS, keyed by the digest that identifies them.
--
-- Review [P1]: the store held 'attempt.input_digest' and
-- 'outputs.manifest_digest' and not one byte of either document. Freeze
-- could not compare a sealed result against the output DECLARATIONS the
-- input manifest names, and intake, publication and restart were left with
-- a number and nothing to replay. Retention is idempotent by construction
-- because the key is computed from the bytes.
CREATE TABLE IF NOT EXISTS manifests (
	digest      TEXT PRIMARY KEY,
	schema      TEXT NOT NULL,
	-- RFC 8785 canonical bytes, so what comes back out is what went in
	body        TEXT NOT NULL,
	retained_at TEXT NOT NULL
) STRICT;

CREATE TABLE IF NOT EXISTS outputs (
	runtime_attempt_id  TEXT PRIMARY KEY
	                    REFERENCES attempts(runtime_attempt_id),
	result_id           TEXT NOT NULL,
	-- the disposition the RESULT declares, which freeze already compared
	-- against the recorded worker_disposition axis
	disposition         TEXT NOT NULL,
	-- the sealed manifest digest this manager RECOMPUTED, not the one the
	-- document declared about itself
	-- the retained result manifest, by the digest this manager recomputed
	manifest_digest     TEXT NOT NULL REFERENCES manifests(digest),
	freeze_operation_id TEXT NOT NULL,
	frozen_at           TEXT NOT NULL
) STRICT;

-- The artifact REFERENCES, one row per declared output. Whether the bytes
-- match is a collection-time fact W2930 owns; that the reference is well
-- formed, digested and free of credentials is decided before it lands here.
CREATE TABLE IF NOT EXISTS output_artifacts (
	runtime_attempt_id TEXT NOT NULL REFERENCES outputs(runtime_attempt_id),
	output_name        TEXT NOT NULL,
	artifact_id        TEXT NOT NULL,
	media_type         TEXT NOT NULL,
	bytes              INTEGER NOT NULL,
	content_digest     TEXT NOT NULL,
	locator            TEXT NOT NULL,
	PRIMARY KEY (runtime_attempt_id, output_name)
) STRICT;
`;

File: v12/python/src/baton_v12/authority/schema.py
"""The v12 Python authority's durable schema.

This is a SELF-CONTAINED store.  It is not the v11 authority, it is not the
Node v12 authority, it is not the Worker Manager control store, and it is not a
migration of any of them.  Nothing here opens `work.sqlite3` or imports
`src/baton_work/`.

What the AUTHORITY owns is fixed by §3 and the table list is the whole
argument: Work state, the per-Work contract selector, the generation counter,
the live assignment, fenced generations, the one typed gate, assignment-ending
and contract events, gate evidence, proposal and workflow receipts, and the
operation journal.  Offers, runtime attempts, quarantined output and runtime
observations are the Worker Manager control store's and are deliberately
absent -- an authority that also stored them would be answering questions it is
not authoritative for.

THE STORE KIND IS WHY THIS SCHEMA IS NOT ENOUGH ON ITS OWN.  The frozen Node
authority also calls its first schema version `1`, and both are SQLite files
holding a table called `meta`.  A version number alone therefore cannot tell
"my store, version 1" from "somebody else's store, version 1", and the failure
mode is not an error -- it is silent adoption.  So every store this module
creates records WHOSE it is before it records how old it is, and the open path
checks the kind first.
"""

__all__ = ["STORE_KIND", "SCHEMA_VERSION", "SCHEMA", "META_STORE_KIND",
           "META_SCHEMA_VERSION", "META_AUTHORITY_UUID"]

# The marker that makes this store recognizably ours.  It names the host, the
# product and the role, because "v12" alone would still be ambiguous between
# the Node authority, this one and the manager's control store.
STORE_KIND = "baton.v12.python.authority"

# W16821 raised this from 1 to 2.  Schema 2 SEPARATES the principal from the
# endpoint address, gives Work an authority-owned effective scope, keys claim
# capacity by principal, and records the authorization decision beside the acts
# it authorized.  A schema-1 store holds none of those facts and nothing in it
# could supply them: approver ruling M33752 makes this a CLEAN INITIALIZATION
# BOUNDARY for disposable proof stores rather than a migration, and
# `store._check_compatibility` refuses the older file read-only and tells the
# operator to initialize a fresh one.
# W29400 raised this from 2 to 3 for the Work-label model: the live set, its
# append-only mutation evidence, and the inverse index the exact filters read.
# Approver ruling M33752 makes each version a CLEAN INITIALIZATION BOUNDARY for
# these disposable proof stores, so this is a version bump and not a migration
# -- which is the schema disposition W29400's own record says must be derived
# at implementation time rather than decided by its decomposition.
# W16823 raised this from 3 to 4, and NO TABLE CHANGED -- which is the whole
# reason the bump has to be argued rather than derived from a diff.  `claim`
# now answers a CLOSED RESULT: the unchanged four-part assignment, the exact
# immutable claim event, and the decision the claim was authorized under.  The
# operation journal retains a committed result whole and replays it byte for
# byte, so a schema-3 journal can hold the OLD BARE ASSIGNMENT and this build
# would hand it to a manager reading it as the new closed document -- a missing
# `decision` discovered at a member lookup rather than at the open.  Versioning
# the RESULT CONTRACT is exactly what the version is for; approver ruling
# M35002 allocates 4 as the cumulative shape after W29400's 3, and no later
# Work may lower or independently reuse it.
#
# W29400, approver ruling M35127: FIVE is the cumulative clean-initialization
# boundary after W16823's 4, and it is allocated ONCE -- here, with the
# creation pipeline it is the boundary for. A Work-label store written under 4
# has no attributable creation act and no `work-create` decision to join a
# create-time label event to, so a build that required them could not adopt
# one; the version is what says so instead of a member lookup discovering it.
#
# The disposition is clean initialization rather than migration, on M33752's
# still-current rule for this disposable proof store: 5 is not reused, not
# lowered, and not reached by upgrading a 4.
SCHEMA_VERSION = 5

META_STORE_KIND = "store_kind"
META_SCHEMA_VERSION = "schema_version"
META_AUTHORITY_UUID = "authority_uuid"

SCHEMA = """
CREATE TABLE IF NOT EXISTS meta (
    key   TEXT PRIMARY KEY,
    value TEXT NOT NULL
) STRICT;

-- Deployment policy.  These are configured facts shared by every Work: which
-- contracts a certified runtime profile can execute, which contract
-- transitions are permitted at all, and the pinned isolation and retention
-- clauses §10.8/§10.9 refer to.
CREATE TABLE IF NOT EXISTS certified_contract (
    contract     TEXT PRIMARY KEY,
    profile      TEXT NOT NULL,
    certified_at TEXT NOT NULL
) STRICT;

CREATE TABLE IF NOT EXISTS contract_transition (
    from_contract TEXT NOT NULL,
    to_contract   TEXT NOT NULL,
    PRIMARY KEY (from_contract, to_contract)
) STRICT;

CREATE TABLE IF NOT EXISTS policy (
    key   TEXT PRIMARY KEY,
    value TEXT NOT NULL
) STRICT;

-- The configuration generation every authorization decision is taken under
-- (§5 of the correction boundary).  ONE ROW, enforced by the primary key
-- check: a table that could hold two generations would have no answer to
-- "which one is current".
--
-- Kept OUT of `policy`, which `set_policy` lets a deployment write freely: a
-- generation a caller could set is a generation a caller could rewind, and an
-- act's recorded provenance would then name a configuration that never
-- existed.
CREATE TABLE IF NOT EXISTS policy_generation (
    one        INTEGER PRIMARY KEY CHECK (one = 1),
    generation INTEGER NOT NULL CHECK (generation >= 1),
    bumped_at  TEXT NOT NULL
) STRICT;

-- W16821.  THE CANONICAL GLOBAL IDENTITY an act is attributed to, separate
-- from the endpoint address it was performed through.
--
-- Two tables rather than a column on `route_handler`, because the mapping is
-- MANY endpoints to ONE principal and that is the whole point: two spellings
-- of one person received two claim slots, and one spelling could not say which
-- scope and which grant authorized an act.
CREATE TABLE IF NOT EXISTS principal (
    principal_id  TEXT PRIMARY KEY,
    registered_at TEXT NOT NULL
) STRICT;

-- The deployment mapping.  The AUTHORITY owns it: a caller names an endpoint
-- and never a principal, so no operand anywhere can choose or widen the
-- identity an act is attributed to.
CREATE TABLE IF NOT EXISTS endpoint (
    participant  TEXT PRIMARY KEY,
    principal_id TEXT NOT NULL,
    bound_at     TEXT NOT NULL,
    FOREIGN KEY (principal_id) REFERENCES principal (principal_id)
) STRICT;

-- One row per Work.  phase is the closed scheduler axis and is NULL exactly
-- when the Work is terminal; gate is the ONE typed gate v11 already displays,
-- never a second axis (§10.7).
CREATE TABLE IF NOT EXISTS work (
    work_id            TEXT PRIMARY KEY,
    route              TEXT NOT NULL,
    status             TEXT NOT NULL,
    phase              TEXT,
    outcome            TEXT,
    rationale          TEXT,
    handler            TEXT,
    contract           TEXT NOT NULL,
    generation_counter INTEGER NOT NULL DEFAULT 0,
    live_generation    INTEGER,
    gate               TEXT,
    -- W16821 item 2.  The effective scope this Work's authorizations are
    -- decided in.  NOT NULL and supplied at creation through the trusted
    -- bootstrap: the correction boundary forbids deriving it from route,
    -- repository or participant spelling, and a nullable column would be a
    -- standing invitation to derive it later "just for the rows that are
    -- missing it".
    scope              TEXT NOT NULL,
    created_at         TEXT NOT NULL
) STRICT;

-- W29400: THE LIVE LABEL SET.  One row per Work per canonical label, so the
-- set is the table rather than a parsed column, and `(work_id, label)` is the
-- uniqueness the contract calls set semantics.
--
-- NO ORDER COLUMN.  A set has no order; the projection sorts, which is what
-- makes two Works with the same labels project identically however they were
-- supplied.
CREATE TABLE IF NOT EXISTS work_label (
    work_id  TEXT NOT NULL,
    label    TEXT NOT NULL,
    added_at TEXT NOT NULL,
    PRIMARY KEY (work_id, label),
    FOREIGN KEY (work_id) REFERENCES work (work_id)
) STRICT;

-- THE INVERSE, because the filters ask "which Work carries this label" and the
-- primary key only answers the other direction.  An exact-membership predicate
-- over a missing index is a scan that gets slower as the deployment grows,
-- and this feature exists to be filtered by.
CREATE INDEX IF NOT EXISTS work_label_by_label
    ON work_label (label, work_id);

-- APPEND-ONLY MUTATION EVIDENCE.  Every EFFECTIVE addition and removal, in
-- order.  A convergent no-op writes nothing here: the contract says adding a
-- present label returns `changed:false` and fabricates no event, because an
-- event that records nothing having happened makes the history unable to say
-- what did.
--
-- The AUTHORIZATION EVIDENCE is not repeated in these columns.  W16821 owns
-- one decision shape in `authorization_decision`, keyed by the act, and this
-- journal names its act rather than carrying a second provisional spelling of
-- endpoint, principal, scope, grant and policy generation -- which is what the
-- parent record asks for in terms.
CREATE TABLE IF NOT EXISTS work_label_event (
    seq     INTEGER PRIMARY KEY AUTOINCREMENT,
    work_id TEXT NOT NULL,
    label   TEXT NOT NULL,
    action  TEXT NOT NULL CHECK (action IN ('added', 'removed')),
    act_id  TEXT NOT NULL,
    at      TEXT NOT NULL
) STRICT;

CREATE TABLE IF NOT EXISTS route_handler (
    route       TEXT NOT NULL,
    participant TEXT NOT NULL,
    PRIMARY KEY (route, participant)
) STRICT;

-- Configured capabilities.  §7 names a distinct actor for the receipt
-- transitions -- verifier, rview, approv, trusted integrator -- and an
-- authorized actor holding the close capability.  Without this, one consumer
-- could publish a candidate, self-verify it, self-review it, self-approve it,
-- integrate it into the canonical target and close the Work, because the
-- receipts stored dispositions and no actor at all.
--
-- A deployment MAY grant one participant several of these: §10.12 says the four
-- receipts stay distinct even when configuration permits one participant to
-- hold more than one role.  What it may not do is leave the question unasked,
-- which is what an actorless receipt does.
--
-- W16821: THE GRANTEE IS THE PRINCIPAL, not an endpoint address.  A grant held
-- by a spelling is a grant a second spelling of the same person does not have,
-- and the deployment could not say so.
--
-- `provenance` admits the whole vocabulary the M6 resolver will need --
-- `direct`, `inherited`, `masked` -- because a durable column that could not
-- hold an inherited grant would have to be migrated to gain one.  What this
-- cut may WRITE is only `direct`, and that restriction lives in the code where
-- a case can remove it and see the suite fail; a CHECK naming one value would
-- have made "the shape admits inheritance" false.
CREATE TABLE IF NOT EXISTS capability (
    principal_id TEXT NOT NULL,
    capability   TEXT NOT NULL,
    scope        TEXT NOT NULL,
    provenance   TEXT NOT NULL CHECK (provenance IN
                   ('direct', 'inherited', 'masked')),
    granted_at   TEXT NOT NULL,
    PRIMARY KEY (principal_id, capability, scope),
    FOREIGN KEY (principal_id) REFERENCES principal (principal_id)
) STRICT;

-- Fenced generations are retained forever with their cause: §10.1 says the
-- counter is never decremented or reused, and the fence is how a late act by an
-- ended generation is refused rather than ignored.
CREATE TABLE IF NOT EXISTS fenced_generation (
    work_id    TEXT NOT NULL,
    generation INTEGER NOT NULL,
    cause      TEXT NOT NULL,
    reason     TEXT,
    fenced_at  TEXT NOT NULL,
    PRIMARY KEY (work_id, generation),
    FOREIGN KEY (work_id) REFERENCES work (work_id)
) STRICT;

-- Deployment-wide capacity (§10.2).  Making this per Work is the bug the table
-- shape forbids -- and W16821 corrects WHOSE capacity it is.
--
-- THE PRINCIPAL IS THE KEY.  One principal holds at most one live claim across
-- the whole deployment, across every endpoint address it acts through.  Keyed
-- by participant, two spellings of one person received two slots, which is a
-- capacity limit that the person it limits can opt out of by being addressed
-- differently.
--
-- The endpoint is kept BESIDE it, not dropped: the Handler, the fence and the
-- assignment identity are all still endpoint-addressed, and a slot that could
-- not say which address took it could not release the right one.
CREATE TABLE IF NOT EXISTS claim_slot (
    principal_id TEXT PRIMARY KEY,
    participant  TEXT NOT NULL,
    work_id      TEXT NOT NULL,
    generation   INTEGER,
    taken_at     TEXT NOT NULL,
    FOREIGN KEY (principal_id) REFERENCES principal (principal_id)
) STRICT;

-- The operation journal.  An identity is durably in exactly one of four states;
-- UNSUBMITTED is the absence of a row (§4).
CREATE TABLE IF NOT EXISTS operation (
    operation_id TEXT PRIMARY KEY,
    signature    TEXT NOT NULL,
    state        TEXT NOT NULL CHECK (state IN ('committed', 'refused', 'retired')),
    result       TEXT,
    detail       TEXT,
    recorded_at  TEXT NOT NULL
) STRICT;

-- W16821 item 5: THE AUTHORIZATION DECISION, IN ONE TABLE.
--
-- The first cut of this correction put four nullable columns on
-- `assignment_event` and three more on `receipt`, and review found what that
-- shape costs: `close` writes neither row, so an authorized close persisted no
-- decision at all, and every act that gained a door would have needed its own
-- copy of the same four columns.  One table, one shape, one writer.
--
-- IMMUTABLE BY ITS PRIMARY KEY.  A decision is what the authority answered at
-- the instant it authorized one act; a second write for the same act would be
-- a second answer to a question that was already decided.
--
-- `act` and `act_id` name the exact act rather than the Work: a claim is keyed
-- by the assignment event it authorized, a close by its Work, each receipt by
-- its own identity, and a durably journalled refused integration attempt by
-- its own row.  Keying a claim by Work would collide the moment a released
-- Work was claimed again.
--
-- ASSIGNMENT-DERIVED ACTS ARE NOT LISTED HERE and that is deliberate: an
-- activity, a contract event and a proposal are carried out UNDER an
-- assignment that was already authorized, so the decision they were performed
-- under is the claim's.  They join to it through the full exact assignment
-- identity rather than copying it, because two copies of one fact are two
-- things that can disagree.
CREATE TABLE IF NOT EXISTS authorization_decision (
    act               TEXT NOT NULL,
    act_id            TEXT NOT NULL,
    endpoint          TEXT NOT NULL,
    principal_id      TEXT NOT NULL,
    effective_scope   TEXT NOT NULL,
    role              TEXT NOT NULL,
    grant_provenance  TEXT NOT NULL CHECK (grant_provenance IN
                        ('direct', 'inherited', 'masked')),
    policy_generation INTEGER NOT NULL,
    decided_at        TEXT NOT NULL,
    PRIMARY KEY (act, act_id),
    FOREIGN KEY (principal_id) REFERENCES principal (principal_id)
) STRICT;

CREATE TABLE IF NOT EXISTS assignment_event (
    seq               INTEGER PRIMARY KEY AUTOINCREMENT,
    work_id           TEXT NOT NULL,
    participant       TEXT NOT NULL,
    generation        INTEGER,
    cause             TEXT NOT NULL,
    fenced            INTEGER NOT NULL,
    reason            TEXT,
    gate              TEXT,
    phase             TEXT,
    at                TEXT NOT NULL
);

-- `claim_seq` IS THE EXACT CLAIM THIS ACT WAS CARRIED OUT UNDER.
--
-- W16821 re-review [P0]: the projection joined back to a claim by
-- `(work_id, participant, generation)` and took the newest match.  That tuple
-- distinguishes v12 claims because they mint generations -- and a v11
-- assignment has NO generation, so releasing and reclaiming through the same
-- endpoint produced two distinct claim acts with identical join fields.  The
-- second claim then became the apparent authorization of the FIRST act's
-- history, without that act or its decision row being touched.
--
-- A nullable tuple, an instant and a newest-row ordering are not an identity.
-- The assignment event's own `seq` is, and it is captured at the moment of the
-- act rather than searched for afterwards.
CREATE TABLE IF NOT EXISTS contract_event (
    seq           INTEGER PRIMARY KEY AUTOINCREMENT,
    work_id       TEXT NOT NULL,
    from_contract TEXT NOT NULL,
    to_contract   TEXT NOT NULL,
    participant   TEXT NOT NULL,
    generation    INTEGER,
    claim_seq     INTEGER NOT NULL,
    rationale     TEXT NOT NULL,
    at            TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS gate_evidence (
    seq      INTEGER PRIMARY KEY AUTOINCREMENT,
    work_id  TEXT NOT NULL,
    gate     TEXT NOT NULL,
    evidence TEXT NOT NULL,
    at       TEXT NOT NULL
);

-- Canonical activity carried out under an exact assignment.  Unique on the
-- assignment plus the caller's key, so a retry is the same row.
CREATE TABLE IF NOT EXISTS activity (
    seq         INTEGER PRIMARY KEY AUTOINCREMENT,
    work_id     TEXT NOT NULL,
    participant TEXT NOT NULL,
    generation  INTEGER,
    claim_seq   INTEGER NOT NULL,
    action_key  TEXT NOT NULL,
    at          TEXT NOT NULL
);

-- IFNULL, not a plain UNIQUE: SQLite treats NULLs as distinct in a unique
-- index, so a v11 assignment (which mints no generation) would insert the same
-- activity twice and the idempotency key would silently stop working.
CREATE UNIQUE INDEX IF NOT EXISTS activity_key
    ON activity (work_id, participant, IFNULL(generation, -1), action_key);

-- The immutable candidate.
--
-- §10.11 requires the receipt to bind the exact assignment AND the input,
-- policy, output, candidate-tree and target digests; §4 adds the frozen result
-- identity and its content digest.  One undifferentiated digest column bound
-- none of that, so a proposal could not say what it was built FROM.  Every
-- digest is required and every one rides the publish operation signature,
-- because later bytes are a new proposal rather than an edit to this one.
CREATE TABLE IF NOT EXISTS proposal (
    proposal_id      TEXT PRIMARY KEY,
    work_id          TEXT NOT NULL,
    participant      TEXT NOT NULL,
    generation       INTEGER,
    claim_seq        INTEGER NOT NULL,
    result_id        TEXT NOT NULL,
    result_digest    TEXT NOT NULL,
    candidate_digest TEXT NOT NULL,
    input_digest     TEXT NOT NULL,
    policy_digest    TEXT NOT NULL,
    target           TEXT NOT NULL,
    published_at     TEXT NOT NULL
) STRICT;

-- The four separately attributable IMMUTABLE receipts (§10.12).
--
-- Rows rather than columns on the proposal, because each carries its own
-- identity, its own actor, and the candidate digest and target revision it was
-- written against.  A receipt that was only a disposition string could not say
-- who attributed it or what they saw.
--
-- kind is UNIQUE per proposal: that is the immutability rule, enforced by the
-- index rather than by a check somebody can forget.
CREATE TABLE IF NOT EXISTS receipt (
    receipt_id        TEXT PRIMARY KEY,
    kind              TEXT NOT NULL CHECK (kind IN
                        ('verification', 'review', 'approval', 'integration')),
    proposal_id       TEXT NOT NULL,
    actor             TEXT NOT NULL,
    disposition       TEXT NOT NULL,
    candidate_digest  TEXT NOT NULL,
    target            TEXT NOT NULL,
    policy_generation INTEGER,
    -- W16821: `actor` stays and is the ENDPOINT that wrote this receipt --
    -- §10.12's separate attributability is about addresses a deployment
    -- configures.  Which principal that address resolved to, in which scope
    -- and by which grant is the DECISION, and it lives in
    -- `authorization_decision` keyed by this receipt's own identity rather
    -- than in three more columns here.
    recorded_at       TEXT NOT NULL,
    FOREIGN KEY (proposal_id) REFERENCES proposal (proposal_id)
) STRICT;

CREATE UNIQUE INDEX IF NOT EXISTS receipt_one_per_kind
    ON receipt (proposal_id, kind);

-- A refused integration is journalled ONCE beside the proposal and never
-- rewrites a committed receipt (§7).
-- W16821 review [P0]: a refused integration attempt is a DURABLE, separately
-- attributable act -- it journals an authorized actor and its operation
-- identity commits -- so it carries the decision that authorized it like any
-- other.  `attempt_id` is what the decision is keyed by; the autoincrement
-- `seq` is an ordering and is not an identity a second table can name
-- stably.
CREATE TABLE IF NOT EXISTS integration_attempt (
    seq         INTEGER PRIMARY KEY AUTOINCREMENT,
    attempt_id  TEXT NOT NULL UNIQUE,
    proposal_id TEXT NOT NULL,
    actor       TEXT NOT NULL,
    reason      TEXT NOT NULL,
    target      TEXT NOT NULL,
    at          TEXT NOT NULL
);
"""
